File: graphs/dijkstra-shortest-path.py
from typing import List

def get_vertex_with_min_distance(distances: List[int], visited: set):
	current_min_distance = float('inf')
	print(distances)
	vertex = None
 
	for vertex_idx, distance in enumerate(distances):
		if vertex_idx in visited:
			continue
		
		if distance <= current_min_distance:
			vertex = vertex_idx
			current_min_distance = distance
	# print(vertex, current_min_distance)
	return vertex, current_min_distance

def dijkstra(start: int, edges: List[int]):
	number_of_vertices = len(edges)
    
	min_distances = [float('inf') for _ in range(number_of_vertices)]

	min_distances[start] = 0
 
	visited = set()

	while len(visited) != number_of_vertices:
		vertex, current_min_distance = get_vertex_with_min_distance(min_distances, visited)
  
		if current_min_distance == float('inf'):
			min_distances = -1
			break
		
		visited.add(vertex)
		
		for edge in edges[vertex]:
			destination, distance_to_desitination = edge

			if destination in visited:
				continue

			new_path_distance = current_min_distance + distance_to_desitination
			current_destination_distance = min_distances[destination]

			if new_path_distance < current_destination_distance:
				min_distances[destination] = new_path_distance
		
	return list(map(lambda x: -1 if x == float('inf') else x, min_distances))

print(
    dijkstra(
    8, 
	[
  [
    [1, 4],
    [7, 11]
  ],
  [
    [0, 4],
    [2, 11],
    [7, 14]
  ],
  [
    [1, 11],
    [3, 10],
    [5, 7],
    [8, 5]
  ],
  [
    [2, 10],
    [4, 12],
    [5, 17]
  ],
  [
    [3, 12],
    [5, 13],
    [6, 3]
  ],
  [
    [2, 7],
    [3, 17],
    [4, 13],
    [6, 5]
  ],
  [
    [4, 3],
    [5, 6],
    [7, 4],
    [9, 8]
  ],
  [
    [0, 11],
    [1, 14],
    [6, 4],
    [8, 10]
  ],
  [
    [2, 5],
    [6, 9],
    [7, 10]
  ],
  []
]
 )
)

File: graphs/dijkstra-shortest-path.js
function getVertexWithMinDistance(minDistances, visited) {
	let vertex = null, currentMinDistance = Infinity

	console.log(minDistances)
	for (let vertexId = 0; vertexId < minDistances.length; vertexId++) {
		const distance = minDistances[vertexId]

		if (visited.includes(vertexId)) continue

		if (distance <= currentMinDistance) {
			vertex = vertexId
			currentMinDistance = distance
		}
	}
	console.log([vertex, currentMinDistance])
	return [vertex, currentMinDistance]
}

function dijkstra(start, edges) {
	let numberOfVertices = edges.length

	let minDistances = new Array(numberOfVertices).fill(Infinity)
	minDistances[start] = 0

	visited = []

	while (visited.length !== numberOfVertices) {
		const [vertex, currentMinDistance] = getVertexWithMinDistance(minDistances, visited)
		console.log(vertex, currentMinDistance)

		visited.push(vertex)

		for (const edge of edges) {
			const [destination, distanceToDestination] = edge
			// if (visited.includes(destination)) continue

			const newPathDistance = currentMinDistance + distanceToDestination
			const currentDestinationDistance = minDistances[destination]

			if (newPathDistance < currentDestinationDistance) {
				minDistances[destination] = newPathDistance
			}
		}

	}

	return minDistances.map((distance) => {
		if (distance === Infinity) return -1
		return distance
	})
}

console.log(
    dijkstra(
    8, 
	[
  [
    [1, 4],
    [7, 11]
  ],
  [
    [0, 4],
    [2, 11],
    [7, 14]
  ],
  [
    [1, 11],
    [3, 10],
    [5, 7],
    [8, 5]
  ],
  [
    [2, 10],
    [4, 12],
    [5, 17]
  ],
  [
    [3, 12],
    [5, 13],
    [6, 3]
  ],
  [
    [2, 7],
    [3, 17],
    [4, 13],
    [6, 5]
  ],
  [
    [4, 3],
    [5, 6],
    [7, 4],
    [9, 8]
  ],
  [
    [0, 11],
    [1, 14],
    [6, 4],
    [8, 10]
  ],
  [
    [2, 5],
    [6, 9],
    [7, 10]
  ],
  []
]
 )
)

